core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -546,9 +546,16 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   def maybeInvokeJoinCallback(member: MemberMetadata,
                               joinGroupResult: JoinGroupResult): Unit = {
     if (member.isAwaitingJoin) {
-      member.awaitingJoinCallback(joinGroupResult)
-      member.awaitingJoinCallback = null
-      numMembersAwaitingJoin -= 1
+      try {
+        member.awaitingJoinCallback(joinGroupResult)
+      } catch {
+        case t: Throwable =>
+          error(s"Failed to invoke join callback for $member due to ${t.getMessage}.")
+          member.awaitingJoinCallback(JoinGroupResult(member.memberId, Errors.UNKNOWN_SERVER_ERROR))
```

Comment:
I guess an alternative is to return a retriable error. We don't know if the failure was a transient issue.

Reply:
Yeah, I considered this. I tend to believe that errors on this path would be fatal errors anyway because they are/would be unexpected errors (likely due to bugs). UNKNOWN_SERVER_ERROR seems quite reasonable.

Reply:
Yeah, I agree it is probably better to fail fast. We'd rather have an explicit error than see the consumer stuck in the rebalance. If we find cases that are transient in the future, we can refine the handling.